### PR TITLE
[Salsa] Acquire d.ts files for JavaScript files

### PR DESCRIPTION
**Pull request by zhengbli**
This PR contains the Salsa feature that acquires `.d.ts` files via tsd to enhance intellisense for JavaScript files. 

More information can be found at the [wiki page](https://github.com/zhengbli/TypeScript/wiki/Salsa)

### CODE REVIEW

File: src/services/jsTyping.ts [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+
+        _host = host;
+        // Clear inferred typings map
+        inferredTypings = {};
+
+        cachePath = cachePath ? cachePath : globalCachePath;
+        // Only infer typings for .js and .jsx files
+        fileNames = fileNames
+            .map(ts.normalizePath)
+            .filter(f => fileExtensionIs(f, ".js") || fileExtensionIs(f, "jsx"));
```
Comment (DanielRosenwasser) [inline src/services/jsTyping.ts]: Use core library functions

File: src/services/jsTyping.ts [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+    interface HostType {
+        directoryExists: (path: string) => boolean;
+        fileExists: (fileName: string) => boolean;
+        readFile: (path: string, encoding?: string) => string;
+        readDirectory: (path: string, extension?: string, exclude?: string[], depth?: number) => string[];
+        writeFile: (path: string, data: string, writeByteOrderMark?: boolean) => void;
+    };
+
+    var _host: HostType;
+    var _safeList: Map<string>;
```
Comment (DanielRosenwasser) [inline src/services/jsTyping.ts]: Turn the linter on for this file

File: src/compiler/commandLineParser.ts [hunk 3]
```diff
@@ -545,6 +546,32 @@ namespace ts {
...
+            const jsonTypingOptions = json["typingOptions"];
+            if (jsonTypingOptions) {
+                for (const id in jsonTypingOptions) {
+                    if (id === "enableAutoDiscovery") {
+                        if (typeof jsonTypingOptions[id] === "boolean") {
+                            options.enableAutoDiscovery = jsonTypingOptions[id];
+                        }
+                    }
+                    else if (id === "include") {
+                        options.include = jsonTypingOptions[id] instanceof Array ? <string[]>jsonTypingOptions[id] : [];
```
Comment (DanielRosenwasser) [inline src/compiler/commandLineParser.ts]: We have an `isArray` function

File: src/services/jsTyping.ts [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+
+            // Remove typings that the user has added to the exclude list
+            for (const excludeTypingName of exclude) {
+                delete inferredTypings[excludeTypingName];
+            }
+
+            const newTypingNames: string[] = [];
+            const cachedTypingPaths: string[] = [];
+            for (const typing in inferredTypings) {
+                if (inferredTypings[typing]) {
```
Comment (DanielRosenwasser) [inline src/services/jsTyping.ts]: It's not clear what you're testing here and why you're doing it. Are you testing for emptiness? Definedness?

File: src/services/jsTyping.ts [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+     * Infer typing names from given file names. For example, the file name "jquery-min.2.3.4.js"
+     * should be inferred to the 'jquery' typing name; and "angular-route.1.2.3.js" should be inferred
+     * to the 'angular-route' typing name.
+     * @param fileNames are the names for source files in the project
+     * @param safeList is the list of names that we are confident they are library names that requires typing
+     */
+    function getTypingNamesFromSourceFileNames(fileNames: string[]) {
+        const jsFileNames = fileNames.filter(f => ts.fileExtensionIs(f, ".js"));
+        const inferredTypingNames = jsFileNames.map(f => ts.removeFileExtension(ts.getBaseFileName(f.toLowerCase())));
+        const cleanedTypingNames = inferredTypingNames.map(f => f.replace(/((?:\.|-)min(?=\.|$))|((?:-|\.)\d+)/g, ""));
```
Comment (DanielRosenwasser) [inline src/services/jsTyping.ts]: Core library functions here

File: src/services/jsTyping.ts [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+     */
+    function getTypingNamesFromNodeModuleFolder(nodeModulesPath: string, filesToWatch: string[]) {
+        // Todo: add support for ModuleResolutionHost too
+        if (!_host.directoryExists(nodeModulesPath)) {
+            return;
+        }
+
+        const typingNames: string[] = [];
+        const packageJsonFiles =
+            _host.readDirectory(nodeModulesPath, /*extension*/undefined, /*exclude*/undefined, /*depth*/2).filter(f => ts.getBaseFileName(f) === "package.json");
```
Comment (DanielRosenwasser) [inline src/services/jsTyping.ts]: Spaces after these comments

File: src/services/jsTyping.ts [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+                for (const newMissingTypingName of newMissingTypingNames) {
+                    notFoundTypingNames.push(newMissingTypingName);
+                }
+            }
+        }
+    }
+
+    function isInstalled(typing: string, installedKeys: string[]) {
+        for (const key of installedKeys) {
+            if (key.indexOf(typing + '/') === 0) {
```
Comment (DanielRosenwasser) [inline src/services/jsTyping.ts]: Move `typing + "/"` into a constant outside of the loop

File: src/services/jsTyping.ts [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+    function getTypingNamesFromCompilerOptions(options: CompilerOptions) {
+        const typingNames: string[] = [];
+        if (!options) {
+            return;
+        }
+
+        if (options.jsx === JsxEmit.React) {
+            typingNames.push("react");
+        }
+        if (options.moduleResolution === ModuleResolutionKind.NodeJs) {
```
Comment (DanielRosenwasser) [inline src/services/jsTyping.ts]: What about CommonJS as a module target?

File: src/services/jsTyping.ts (hunk 4 above)
Comment (zhengbli) [inline src/services/jsTyping.ts]: It is to test if the value is `undefined` or not.
So the `inferredTypings` is a map from the name of the library, e.g. jquery, commander to the actual `.d.ts` file path. Initially when we inferred that we need the `.d.ts` files of a library, we add a key to the `inferredTypings` map, with a value of `undefined`. Then after all the searching, if the key has a new value, that means we found the `.d.ts` files in the local cache, therefore the value (actual `.d.ts` file path should be added to the `cachedTypingPaths` to return later); if not, the name of the library should be added to the `newTypingNames` so we can tell tsd to download it.

File: src/services/jsTyping.ts (hunk 8 above)
Comment (zhengbli) [inline src/services/jsTyping.ts]: Would that be a strong enough evidence of using Node? @billti has commented before with the concern that the Common-JS style is used in many places outside the Node environment:
> Lots of folks use CommonJS-style modules and then a bundler (browserify, webpack, etc.) for web development, and aren’t necessarily doing Node development

File: src/compiler/commandLineParser.ts [hunk 9]
```diff
@@ -545,6 +546,32 @@ namespace ts {
...
+                    }
+                    else if (id === "exclude") {
+                        options.exclude = jsonTypingOptions[id] instanceof Array ? <string[]>jsonTypingOptions[id] : [];
+                    }
+                    else {
+                        // Todo: change diagnostics
```
Comment (billti) [inline src/compiler/commandLineParser.ts]: I don't understand what this TODO means? If you are not going to do it this code review, please make any it clearer for anyone else.

File: src/compiler/sys.ts [hunk 10]
```diff
@@ -150,12 +150,17 @@ namespace ts {
...
+            function readDirectory(path: string, extension?: string, exclude?: string[], depth?: number): string[] {
                 const result: string[] = [];
                 exclude = map(exclude, s => getCanonicalPath(combinePaths(path, s)));
-                visitDirectory(path);
+                visitDirectory(path, depth);
                 return result;
-                function visitDirectory(path: string) {
+
+                function visitDirectory(path: string, depth?: number) {
+                    if (depth <= 0) {
```
Comment (billti) [inline src/compiler/sys.ts]: This depends on the funky behavior that "undefined <= 0" is false in order not to break when the optional argument isn't provided. Can you check the argument is provided before checking it is 0 or less (e.g. "depth !== undefined && ..").

File: src/compiler/sys.ts [hunk 11]
```diff
@@ -16,7 +16,7 @@ namespace ts {
         createDirectory(path: string): void;
         getExecutingFilePath(): string;
         getCurrentDirectory(): string;
-        readDirectory(path: string, extension?: string, exclude?: string[]): string[];
+        readDirectory(path: string, extension?: string, exclude?: string[], depth?: number): string[];
```
Comment (billti) [inline src/compiler/sys.ts]: Does the ChakraHost.readDirectory implementation honor this new parameter you added? For ChakraHost we simply delegate to the host implementation or readDirectory.

File: src/compiler/types.ts [hunk 12]
```diff
@@ -1590,7 +1590,7 @@ namespace ts {
     }
 
     export interface WriteFileCallback {
-        (fileName: string, data: string, writeByteOrderMark: boolean, onError?: (message: string) => void): void;
+        (fileName: string, data: string, writeByteOrderMark?: boolean, onError?: (message: string) => void): void;
```
Comment (billti) [inline src/compiler/types.ts]: Why did this change to optional?

File: src/server/editorServices.ts [hunk 13]
```diff
@@ -16,6 +16,11 @@ namespace ts.server {
     }
 
     const lineCollectionCapacity = 4;
+    const getCanonicalFileName = createGetCanonicalFileName(sys.useCaseSensitiveFileNames);
+    const globalCachePath = toPath(
+        ".typingCache",
+        process.env[(process.platform === "win32") ? "USERPROFILE" : "HOME"],
```
Comment (billti) [inline src/server/editorServices.ts]: This is environment specific, as it depends on the global `process` being available, which is a Node.js thing. Please abstract this (probably via `ServerHost` which uses `System` from `sys.ts`).

File: src/server/editorServices.ts [hunk 14]
```diff
@@ -783,6 +789,10 @@ namespace ts.server {
                     this.openFileRoots = openFileRoots;
                     this.openFileRoots.push(info);
                 }
+
+                if (fileExtensionIs(info.fileName, ".js") || fileExtensionIs(info.fileName, ".jsx")) {
```
Comment (billti) [inline src/server/editorServices.ts]: So you can't set `autoAquireTypings` to `true` for TypeScript files?

File: src/server/editorServices.ts [hunk 15]
```diff
@@ -1037,6 +1053,97 @@ namespace ts.server {
             return undefined;
         }
 
+        acquireTypingForJs(project: Project) {
+            if (!project) { return; }
+
+            let tsd: any;
+            try {
+                tsd = require("tsd");
```
Comment (billti) [inline src/server/editorServices.ts]: Node.js specific code outside of `sys.ts`.

File: src/server/editorServices.ts [hunk 16]
```diff
@@ -1037,6 +1053,97 @@ namespace ts.server {
...
+            const { cachedTypingPaths, newTypingNames } = ts.JsTyping.discoverTypings(
+                sys,
+                project.getFileNames(),
+                globalCachePath,
+                cachePath,
+                typingOptions,
+                compilerOptions
+            );
+
+            // Bail out when the autoTyping is disabled
```
Comment (billti) [inline src/server/editorServices.ts]: I don't see how this comment matches the logic below.

File: src/server/editorServices.ts [hunk 17]
```diff
@@ -1037,6 +1053,97 @@ namespace ts.server {
...
+
+                promise
+                    .then(() => api.select(query, options))
+                    .then((selection: any) => api.install(selection, options))
+                    .then((installResult: any) => {
+                        const keys = Object.keys(installResult.written.dict);
+                        keys.forEach(key => addTypingToProject(ts.combinePaths(typingsPath, key), project));
+                        project.projectService.updateProjectStructure();
+                        if (cachePath !== globalCachePath) {
+                            JsTyping.updateNotFoundTypingNames(newTypingNames, cachePath);
```
Comment (billti) [inline src/server/editorServices.ts]: Why don't we also track not found names in the global cache to avoid retrying repeatedly there? (Add a comment to explain, as it's non-obvious).

File: src/server/editorServices.ts [hunk 18]
```diff
@@ -1037,6 +1053,97 @@ namespace ts.server {
...
+                        }
+                    });
```
Comment (billti) [inline src/server/editorServices.ts]: It would be good to add a failure handler on the end of the promise chain, even if it just logs.

File: src/server/editorServices.ts (hunk 14 above)
Comment (zhengbli) [inline src/server/editorServices.ts]: This line is only for inferred project that doesn't have a config file, namely loose files. So currently I am not inferring typings for loose `.ts` files. For projects with config file, one can set typing options in the config file for either `tsconfig.json` or `jsconfig.json`, though all the inferring still only happen on `.js` files

File: src/server/editorServices.ts (hunk 15 above)
Comment (zhengbli) [inline src/server/editorServices.ts]: This is not `sys.ts` though?

File: src/services/shims.ts [hunk 19]
```diff
@@ -410,7 +412,7 @@ namespace ts {
             }
         }
 
-        public readDirectory(rootDir: string, extension: string, exclude: string[]): string[] {
+        public readDirectory(rootDir: string, extension: string, exclude: string[], depth?: number): string[] {
```
Comment (billti) [inline src/services/shims.ts]: So is this parameter just not used in the function? That doesn't seem right.

File: src/server/editorServices.ts (hunk 16 above)
Comment (zhengbli) [inline src/server/editorServices.ts]: I was trying to say "Bail out when no actions needed", will update to make it clearer.

File: src/services/shims.ts (hunk 19 above)
Comment (zhengbli) [inline src/services/shims.ts]: Oops, the implementation was override during merging with master. Will fix soon.

File: src/server/editorServices.ts (hunk 15 above)
Comment (zhengbli) [inline src/server/editorServices.ts]: Oh I guess you meant the same with the `process.env.HOME` line above. Got it.

File: src/services/jsTyping.ts [hunk 20]
```diff
@@ -0,0 +1,282 @@
...
+/* @internal */
+namespace ts.JsTyping {
+
+    interface HostType {
+        directoryExists: (path: string) => boolean;
+        fileExists: (fileName: string) => boolean;
+        readFile: (path: string, encoding?: string) => string;
+        readDirectory: (path: string, extension?: string, exclude?: string[], depth?: number) => string[];
+        writeFile: (path: string, data: string, writeByteOrderMark?: boolean) => void;
+    };
```
Comment (billti) [inline src/services/jsTyping.ts]: Why are you declaring the above interface, which is a subset of `System`, rather than just using `System`? Isn't the type of arg passed a `System` anyway?

File: src/services/jsTyping.ts [hunk 21]
```diff
@@ -0,0 +1,282 @@
...
+     * @param compilerOptions is used for typing inferring. 
+     */
+    export function discoverTypings(
+        host: HostType,
+        fileNames: string[],
+        globalCachePath: Path,
+        cachePath: Path,
+        typingOptions: TypingOptions,
+        compilerOptions?: CompilerOptions)
+        : { cachedTypingPaths: string[], newTypingNames: string[], filesToWatch: string[], newTsdJsonPath?: string } {
```
Comment (billti) [inline src/services/jsTyping.ts]: The `@param` names and order don't match the actual parameters on the function (see `@projectRootPath` and `host`).

File: src/services/jsTyping.ts [hunk 22]
```diff
@@ -0,0 +1,282 @@
...
+
+    interface HostType {
+        directoryExists: (path: string) => boolean;
+        fileExists: (fileName: string) => boolean;
+        readFile: (path: string, encoding?: string) => string;
+        readDirectory: (path: string, extension?: string, exclude?: string[], depth?: number) => string[];
+        writeFile: (path: string, data: string, writeByteOrderMark?: boolean) => void;
+    };
+
+    var _host: HostType;
```
Comment (billti) [inline src/services/jsTyping.ts]: This is dangerously structured. You declare `_host` here, and use it in several top level functions in the namespace, but it's not set until the exported function `discoverTypings` is called. While those other use sites may only be called after/from `discoverTypings` currently, this may not be obvious to someone else adding or exporting more functions. Either initialize it when the namespace is initialized, or pass it from `discoverTypings` to the functions it calls that need it.

File: src/services/jsTyping.ts [hunk 23]
```diff
@@ -0,0 +1,282 @@
...
+        typingOptions: TypingOptions,
+        compilerOptions?: CompilerOptions)
+        : { cachedTypingPaths: string[], newTypingNames: string[], filesToWatch: string[], newTsdJsonPath?: string } {
+
+        _host = host;
+        // Clear inferred typings map
+        inferredTypings = {};
+
+        cachePath = cachePath ? cachePath : globalCachePath;
+        // Only infer typings for .js and .jsx files
```
Comment (billti) [inline src/services/jsTyping.ts]: Again, not sure we want to restrict this to only projects containing JavaScript files. If I have a project with TypeScript files, and I set `AutoAcquireTypings` to `true`, shouldn't it also get the typings for my Node modules as it would a project with JavaScript files?

We also don't know which files are using which libraries, so if I have a ton of code in TypeScript using libraries and modules, it won't get any typings automatically, but even if I add one JavaScript file (even if it's empty), we'll go get all the typings for your project?? That seems a little weird right?

File: src/services/jsTyping.ts [hunk 24]
```diff
@@ -0,0 +1,282 @@
...
+                    cachedTypingPaths.push(inferredTypings[typing]);
+                }
+                else {
+                    newTypingNames.push(typing);
+                }
+            }
+            return { cachedTypingPaths, newTypingNames, filesToWatch, newTsdJsonPath };
+        }
+        else {
+            return { cachedTypingPaths: [], newTypingNames: [], filesToWatch: [], newTsdJsonPath: null }
```
Comment (billti) [inline src/services/jsTyping.ts]: Why not just do this on function entry `if(!typingOptions)`, and skip some of the other processing that isn't needed (e.g. iterating over the filenames, checking for the safeList file, etc.)?

File: src/services/jsTyping.ts [hunk 25]
```diff
@@ -0,0 +1,282 @@
...
+        readFile: (path: string, encoding?: string) => string;
+        readDirectory: (path: string, extension?: string, exclude?: string[], depth?: number) => string[];
+        writeFile: (path: string, data: string, writeByteOrderMark?: boolean) => void;
+    };
+
+    var _host: HostType;
+    var _safeList: Map<string>;
+
+    // a typing name to typing file path mapping
+    var inferredTypings: Map<string> = {};
```
Comment (billti) [inline src/services/jsTyping.ts]: This is dangerously structured also. You are storing this value on the namespace (i.e. a singleton per server/language service) and accessing it from several top-level functions, yet the entries it contains are specific to a project.

File: src/services/jsTyping.ts (hunk 20 above)
Comment (zhengbli) [inline src/services/jsTyping.ts]: This is because this code is shared with VS, and on the VS side they use `CoreServicesShimHostAdapter` instead of `System`.

File: src/services/jsTyping.ts (hunk 8 above)
Comment (billti) [inline src/services/jsTyping.ts]: Regardless, lots of people use Node module resolution (e.g. Angular2 development) or CommonJS module format (e.g. Browserify or WebPack) when they are not writing code for Node.js.

Including node.d.ts does pollute the global scope somewhat (with objects list `process`, `buffer`, etc.), so I probably wouldn't include it just for specifying `CommonJS` as the module kind (which is quite common). Leaving it for `NodeJS` resolution seems fine though, as then you are more likely to be using NodeJS.

File: src/services/jsTyping.ts [hunk 26]
```diff
@@ -0,0 +1,282 @@
...
+    }
+
+    function getTypingNamesFromCompilerOptions(options: CompilerOptions) {
+        const typingNames: string[] = [];
+        if (!options) {
+            return;
+        }
+
+        if (options.jsx === JsxEmit.React) {
+            typingNames.push("react");
```
Comment (billti) [inline src/services/jsTyping.ts]: I think this would be better detected based on file extension (i.e. if writing any `.jsx` or `.tsx` files). If you are writing JavaScript, you are unlikely to be settings TypeScript compiler options like the JSX emit type.

File: src/services/jsTyping.ts (hunk 25 above)
Comment (zhengbli) [inline src/services/jsTyping.ts]: The `inferredTypings` is cleared every time a new inferring process is started (the `discoverTypings` is called), so the information it stores is not used either globally or by a project. Maybe I should clear the object at the end of the `discoverTypings` function to save memory, though that's not a structure issue.

File: src/services/jsTyping.ts (hunk 24 above)
Comment (zhengbli) [inline src/services/jsTyping.ts]: Good point. It makes sense to quit earlier.

File: src/compiler/sys.ts (hunk 11 above)
Comment (jramsay) [inline src/compiler/sys.ts]: We've updated this on the VS host side as well.  I spoke to @mhegazy and @vladima -  we'll need to guard against this change in shims.ts:
 public readDirectory(rootDir: string, extension: string, exclude: string[], depth?: number): string[] {
            // Wrap the API changes for 1.8 release. This try/catch
            // should be removed once TypeScript 1.8 has shipped.
           let encoded: string;
            try {
                encoded = this.shimHost.readDirectory(rootDir, extension, JSON.stringify(exclude), depth);
            }
            catch (e) {
                encoded = this.shimHost.readDirectory(rootDir, extension, JSON.stringify(exclude));
            }
            return JSON.parse(encoded);
        }

File: src/compiler/types.ts (hunk 12 above)
Comment (jramsay) [inline src/compiler/types.ts]: We should be able to revert this change - I don't think it is required.

File: src/server/editorServices.ts (hunk 17 above)
Comment (zhengbli) [inline src/server/editorServices.ts]: The original thought is that inferred projects have very few project level activities compared to configured projects (add/remove files, open/close files), therefore when something is not found for a inferred project, it would only impact the first time, namely upon project creation. So in the offline case, We don't need to wait until the end of the session to reset the "not found" list. Though on a second thought the gain is probably trivial, as it would interfere with other opened configured projects anyway. It might be easier just to do the same thing for inferred projects.

**Comment by zhengbli** [comment]
@vladima @mhegazy do you have comments? Thanks

File: src/services/jsTyping.ts [hunk 27]
```diff
@@ -0,0 +1,271 @@
...
+     * that we won't try again next time within thie session.
+     * @param cachedTypingsPaths The list of resolved cached d.ts paths
+     * @param newTypings The list of new typings that the host attempted to acquire using TSD
+     * @param cachePath The path to the local tsd.json cache
+     */
+    export function updateNotFoundTypingNames(triedTypingNames: string[], cachePath: string, host: HostType): void {
+        const tsdJsonPath = ts.combinePaths(cachePath, "tsd.json");
+        const cacheTsdJsonDict = tryParseJson(tsdJsonPath, host);
+        if (cacheTsdJsonDict) {
+            if (cacheTsdJsonDict.hasOwnProperty("installed")) {
```
Comment (jramsay) [inline src/services/jsTyping.ts]: We always need to check for missing typings even when there aren't any installed. (I hit this case for the Cordova project).  Here is how I fixed it locally:

 let installedTypingFiles: string[] = [];
        if (cacheTsdJsonDict) {
            if (cacheTsdJsonDict.hasOwnProperty("installed")) {
                installedTypingFiles = Object.keys(cacheTsdJsonDict.installed);
            }

Comment (zhengbli) [inline src/services/jsTyping.ts]: I didn't see the differences?

File: src/services/jsTyping.ts [hunk 28]
```diff
@@ -0,0 +1,271 @@
...
+        const filesToWatch: string[] = [];
+        // Directories to search for package.json, bower.json and other typing information
+        let searchDirs: string[] = [];
+        let exclude: string[] = [];
+
+        mergeTypings(typingOptions.include);
+        exclude = typingOptions.exclude ? typingOptions.exclude : [];
+
+        if (typingOptions.enableAutoDiscovery) {
+            searchDirs = ts.deduplicate(fileNames.map(ts.getDirectoryPath));
```
Comment (jramsay) [inline src/services/jsTyping.ts]: search directory should also include the projectRootPath if it is set. There are cases when there are no files in the root directory but the package.json and bower.json are located there.

File: src/services/jsTyping.ts [hunk 29]
```diff
@@ -0,0 +1,271 @@
...
+     * @param globalCachePath is used 1. to get the safe list file path 2. as cachePath if it is not specified.
+     * @param cachePath is the actual path to get the cached d.ts files from.
+     * @param typingOptions are used for customizing the typing inferring process.
+     * @param compilerOptions are used as a source of typing inferring.
+     */
+    export function discoverTypings(
+        host: HostType,
+        fileNames: string[],
+        globalCachePath: Path,
+        cachePath: Path,
```
Comment (jramsay) [inline src/services/jsTyping.ts]: rename cachePath to projectRootPath?

File: src/services/jsTyping.ts [hunk 30]
```diff
@@ -0,0 +1,271 @@
...
+        : { cachedTypingPaths: string[], newTypingNames: string[], filesToWatch: string[] } {
+
+        // a typing name to typing file path mapping
+        const inferredTypings: Map<string> = {};
+
+        if (!typingOptions) {
+            return { cachedTypingPaths: [], newTypingNames: [], filesToWatch: [] };
+        }
+
+        cachePath = cachePath ? cachePath : globalCachePath;
```
Comment (jramsay) [inline src/services/jsTyping.ts]: const cachePath = projectRootPath ? projectRootPath : globalCachePath;

**Comment by zhengbli** [comment]
Close this in favor of #7179
